src/casts/specify-representation.rkt

```diff
@@ -1236,90 +1213,112 @@ but a static single assignment is implicitly maintained.
+                   (assign$ a-t1 t1)
+                   (assign$ a-t2 t2)
+                   (invoke-cast-tuple a-v a-t1 a-t2))])]
         [(Make-Tuple-Coercion mk-crcn (app recur t1) (app recur t2) (app recur l))
-         (: invoke-mk-tuple-crcn ((Code-Label Uid) (Var Uid) D0-Expr D0-Expr -> D0-Expr))
-         (define (invoke-mk-tuple-crcn mk-tuple-crcn t1 t2 l)
-           (App-Code
-            mk-tuple-crcn
-            (list t1 t2 l ZERO-IMDT
-                  (sr-tagged-array-ref t2 TYPE-TUPLE-TAG TYPE-TUPLE-COUNT-INDEX))))
-         (let ([mk-tuple-crcn (get-mk-tuple-crcn! mk-crcn)])
-           (if (Var? t1)
-               (invoke-mk-tuple-crcn mk-tuple-crcn t1 t2 l)
-               (begin$
-                 (assign$ tuple-type1 t1)
-                 (invoke-mk-tuple-crcn mk-tuple-crcn tuple-type1 t2 l))))]
-        [(Compose-Tuple-Coercion compose (app recur c1) (app recur c2))
-         (: invoke-comp ((Code-Label Uid) (Var Uid) D0-Expr -> D0-Expr))
-         (define (invoke-comp comp-tuple c1 c2)
-           (begin$
-             (assign$ tagged-count
-               (sr-tagged-array-ref
-                c2 COERCION-MEDIATING-TAG COERCION-TUPLE-COUNT-INDEX))
-             (assign$ count (op$ %>> tagged-count COERCION-SECOND-TAG-SHIFT))
-             (App-Code comp-tuple (list c1 c2 ZERO-IMDT count TRUE-IMDT))))
-         (let ([mk-tuple-crcn (get-comp-tuple-crcn! compose)])
-           (if (Var? c1)
-               (invoke-comp mk-tuple-crcn c1 c2)
-               (begin$
-                 (assign$ tuple-crcn1 c1)
-                 (invoke-comp mk-tuple-crcn tuple-crcn1 c2))))]
+         (define mk-tuple-crcn (get-mk-tuple-crcn! mk-crcn))
+         (: invoke-mk-tuple-crcn ((Var Uid) (Var Uid) -> D0-Expr))
+         (define (invoke-mk-tuple-crcn t1 t2) (app-code$ mk-tuple-crcn t1 t2 l))
+         (cond
+           [(and (Var? t1) (Var? t2)) (invoke-mk-tuple-crcn t1 t2)]
+           [(Var? t1) (begin$ (assign$ a-t2 t2) (invoke-mk-tuple-crcn t1 a-t2))]
+           [(Var? t2) (begin$ (assign$ a-t1 t1) (invoke-mk-tuple-crcn a-t1 t2))]
+           [else (begin$
+                   (assign$ a-t1 t1)
+                   (assign$ a-t2 t2)
+                   (invoke-mk-tuple-crcn a-t1 a-t2))])]
+        [(Compose-Tuple-Coercion comp-crcn (app recur c1) (app recur c2))
+         (define comp-tuple-crcn (get-comp-tuple-crcn! comp-crcn))
+         (: invoke-comp-tuple-crcn ((Var Uid) (Var Uid) -> D0-Expr))
+         (define (invoke-comp-tuple-crcn c1 c2)
+           (app-code$ comp-tuple-crcn c1 c2))
+         (cond
+           [(and (Var? c1) (Var? c2)) (invoke-comp-tuple-crcn c1 c2)]
+           [(Var? c1) (begin$ (assign$ a-c2 c2) (invoke-comp-tuple-crcn c1 a-c2))]
+           [(Var? c2) (begin$ (assign$ a-c1 c1) (invoke-comp-tuple-crcn a-c1 c2))]
+           [else (begin$
+                   (assign$ a-c1 c1)
+                   (assign$ a-c2 c2)
+                   (invoke-comp-tuple-crcn a-c1 a-c2))])]
```

Comment:
This chunk of code, that is doing manual ANF, doesn't seem to be necessary.

Reply:
Done.